### PR TITLE
Implement justify interval methods

### PR DESCRIPTION
**Pull request by jkosh44**
Postgres defines the justify_days, justify_hours, and justify_interval
functions, which explicitly convert between months, days, and
microseconds within an Interval. This commit adds these functions.

### Motivation
This PR adds a feature that has not yet been specified. Due to a limitation in Materialize, we diverge from PostgreSQL when comparing Intervals using equality and ordering. Specifically we do not combine months, days, and microseconds like PostgreSQL does. These functions will allow users to explicitly opt into this functionality by using the justify methods.

### Testing

- [X] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - Add the \[\`justify_days\`\](/sql/functions/justify-days), \[\`justify_hours\`\](/sql/functions/justify-hours), and \[\`justify_interval\`\](/sql/functions/justify-interval) functions.

### CODE REVIEW

File: src/repr/src/adt/interval.rs [hunk 1]
```diff
@@ -551,6 +567,98 @@ impl Interval {
 
         Some(res)
     }
+
+    /// Adjust interval so 'days' contains less than 30 days, adding the excess to 'months'.
```
Comment (jkosh44) [inline src/repr/src/adt/interval.rs:571]: Most of this logic comes directly from Postgres: https://github.com/postgres/postgres/blob/73c61a50a1555007001d29844dcdb10b4f982a73/src/backend/utils/adt/timestamp.c#L2695-L2825

And an approved patch that removes some overflow issues from the postgres code: https://commitfest.postgresql.org/37/3540/

**Review by sploiselle — COMMENTED** [review]
We should move these impls to the "new" function framework. You can find an example of what that looks like here:
https://github.com/MaterializeInc/materialize/pull/9513/files

The big catch with this is that you have to remember to add all of the functions to the `derive_unary` macro.

You can move these implementations into `src/expr/src/scalar/func/impls/interval.rs`, but the "actual" implementations can stay in the `adt` crate. If that's unclear/ambiguous, lmk.

File: test/sqllogictest/interval.slt [hunk 2]
```diff
@@ -1063,3 +1063,72 @@ query T
 SELECT INTERVAL '-1 day' < INTERVAL '-9999999 hours';
 ----
 true
+
+# Justify tests
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply test/sqllogictest/interval.slt:1067]: These are all 3(!!) of the PG justify tests: https://github.com/postgres/postgres/blob/73c61a50a1555007001d29844dcdb10b4f982a73/src/test/regress/expected/interval.out#L386-L404

The rest come from the patch I linked to earlier:  https://commitfest.postgresql.org/37/3540/

File: src/repr/src/adt/interval.rs [hunk 3]
```diff
@@ -551,6 +567,98 @@ impl Interval {
...
+
+        Self::new(months, days, self.micros)
+    }
+
+    fn justify_days_inner(months: i32, days: i32) -> Result<(i32, i32), anyhow::Error> {
+        let days_per_month = i32::from(Self::DAY_PER_MONTH);
+        let whole_month = days / days_per_month;
+        let days = days - whole_month * days_per_month;
+
+        let months = match months.checked_add(whole_month) {
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/repr/src/adt/interval.rs]: Thanks! I was trying to find a more concise way of doing this and couldn't figure it out. This is much nicer. I've switched to this syntax.

File: src/repr/src/adt/interval.rs [hunk 4]
```diff
@@ -551,6 +567,98 @@ impl Interval {
...
+        }
+
+        Self::new(self.months, days, micros)
+    }
+
+    fn justify_hours_inner(days: i32, micros: i64) -> Result<(i32, i64), anyhow::Error> {
+        let days = match i32::try_from(micros / &*USECS_PER_DAY)
+            .ok()
+            .and_then(|d| days.checked_add(d))
+        {
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/repr/src/adt/interval.rs]: Fixed

File: src/repr/src/adt/interval.rs [hunk 5]
```diff
@@ -551,6 +567,98 @@ impl Interval {
...
+
+    /// Adjust interval so 'days' contains less than 30 days, adding the excess to 'months'.
+    /// Adjust interval so 'micros' contains less than a whole day, adding the excess to 'days'.
+    /// Also, the sign bit on all three fields is made equal, so either all three fields are negative or all are positive.
+    pub fn justify_interval(&self) -> Result<Self, anyhow::Error> {
+        let days_per_month = i32::from(Self::DAY_PER_MONTH);
+        let mut months = self.months;
+        let mut days = self.days;
+        let micros = self.micros;
+        // We justify days twice to avoid try an intermediate overflow of days if it would be able
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/repr/src/adt/interval.rs]: I think I meant "to try to avoid"

**Review by sploiselle — APPROVED** [review]

**Pull request merged**